src/browser/fake_network.js

```diff
@@ -43,6 +84,179 @@ function iptolong(parts) {
+                const buf_len_left = new_tail % capacity;
+                const buf_len_right = capacity - this.tail;
+                dst_array.set(buffer.subarray(this.tail));
+                dst_array.set(buffer.subarray(0, buf_len_left), buf_len_right);
+            }
+            else {
+                dst_array.set(buffer.subarray(this.tail, new_tail));
+            }
+        }
+        return length;
+    }
+
+    /**
+     * @param {number} length
+     */
+    remove(length)
+    {
+        if(length > this.length) {
+            length = this.length;
+        }
+        if(length) {
+            this.tail = (this.tail + length) % this.buffer.length;
+            this.length -= length;
+        }
+        return length;
+    }
+}
+
+function create_eth_encoder_buf()
+{
+    const eth_frame = new Uint8Array(ETH_FRAME_SIZE);
+    const buffer = eth_frame.buffer;
+    const offset = eth_frame.byteOffset;
+    return {
+        eth_frame: eth_frame,
+        eth_frame_view: new DataView(buffer),
+        eth_payload_view: new DataView(buffer, offset + ETH_PAYLOAD_OFFSET, ETH_PAYLOAD_SIZE),
+        ipv4_payload_view: new DataView(buffer, offset + IPV4_PAYLOAD_OFFSET, IPV4_PAYLOAD_SIZE),
+        udp_payload_view: new DataView(buffer, offset + UDP_PAYLOAD_OFFSET, UDP_PAYLOAD_SIZE),
+        text_encoder: new TextEncoder()
+    };
+}
+
+/**
+ * Copy given data array into view starting at offset, return number of bytes written.
+ *
+ * @param {number} offset
+ * @param {ArrayBuffer|ArrayBufferView} data
+ * @param {DataView} view
+ * @param {Object} out
+ */
+function view_setArray(offset, data, view, out)
+{
+    out.eth_frame.set(data, view.byteOffset + offset);
+    return data.length;
+}
+
+/**
+ * UTF8-encode given string into view starting at offset, return number of bytes written.
+ *
+ * @param {number} offset
+ * @param {string} str
+ * @param {DataView} view
+ * @param {Object} out
+ */
+function view_setString(offset, str, view, out)
+{
+    const ofs = view.byteOffset + offset;
+    const result = out.text_encoder.encodeInto(str, ofs ? out.eth_frame.subarray(ofs) : out.eth_frame);
+    return result.written;
+}
+
+/**
+ * Calculate internet checksum for view[0 : length], store 16-bit result in view at offset.
+ * Source: RFC768 and RFC1071 (chapter 4.1).
+ *
+ * @param {number} offset
+ * @param {number} length
+ * @param {number} checksum
+ * @param {DataView} view
+ * @param {Object} out
+ */
+function view_setInetChecksum(offset, length, checksum, view, out)
+{
+    const uint16_end = view.byteOffset + (length & ~1);
+    const eth_frame = out.eth_frame;
+    for(let i = view.byteOffset; i < uint16_end; i += 2) {
+        checksum += eth_frame[i] << 8 | eth_frame[i+1];
+    }
+    if(length & 1) {
+        checksum += eth_frame[uint16_end] << 8;
+    }
+    while(checksum >> 16) {
+        checksum = (checksum & 0xffff) + (checksum >> 16);
+    }
+    view.setUint16(offset, ~checksum);
+}
```

Comment:
I'd suggest renaming this to `calc_inet_checksum` (or similar), returning the checksum and doing the `view.setUint16` in the caller.

Reply:
Good idea, done.